### ISSUE TITLE
Array type checking not working with functions returning any[]

### ISSUE THREAD

**Issue by mhchem** [Microsoft/TypeScript#20238]
**TypeScript Version:**  2.6.1

**Code**

```ts
type SomethingSpecial = "abc";
function anyArray(): any[] { return [2]; }

var a: SomethingSpecial[];
a = [2];  // error
a = anyArray();  // no error
```

**Expected behavior:**
The last two lines should behave identical with regards to error checking.

**Actual behavior:**
The last line currently does not give an error.

**Comment by Proculopsis** [comment]
I'm not sure it makes sense to have an array of type of literal string ("abc").

It's no longer an issue with:

type SomethingSpecial = number; // "abc";

It looks like the transpiler's behaviour is expected.

**Comment by Proculopsis** [comment]
The results are predictable for:

type SomethingSpecial =  "abc" | "xyz";

var a: SomethingSpecial[];
a = ["abc"];  // no error
a = ["abc","xyz"];  // no error
a = ["pqr"];  // error

**Comment by gcnew** [comment]
This is the expected and correct behaviour. The `any` type is assignable to everything and from anything.

If that were not your intention, just remove the explicit `any[]` return type and `number[]` will be inferred, producing an error as you expect.

PS: `any` is really dangerous and contaminating. Using it even once you assume responsibility for all code paths it may end up in.

**Comment by mhchem** [comment]
First of all, my example was of course a minimal example that demonstrates the issue with least amount of code. The real code where this behavior was observed, is waaay more complex, with assignments to arrays of discriminated unions.

Ooops! I can assign `any` to any type? I didn't expect that. I knew that I can assign any value to an `any` varible. But I find it strange to be the other way round.

Okay, the issue now changed to: **Documentation needs to be updated.** https://www.typescriptlang.org/docs/handbook/basic-types.html#any just talks about assignments *to* any. It should also contain a big warning about assignments *from* any.

As a single `any` can spoil my whole code (TypeScript making it look to be type-save, but still containing wrong-type errors), is here a way to disallow assignments from `any` to anything more specific?

**Comment by gcnew** [comment]
> is here a way to disallow assignments from any to anything more specific?

Yes, use `{} | null | undefined` instead.

```ts
let x: {} | null | undefined;

x = 5; // OK
x = "tomato"; // OK

let y: number = x; // Error
let z: string[] = x; // Error
```
